### PR TITLE
JavaScript: Keep unary expressions parentheses with comments

### LINKED ISSUE #4154: Ugly print for !! with comment
**Prettier 1.11.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEqAFAJT7AA6U++ATnDOrdQIQtFU00D0AVPpAC2ghDHy9unLqggQpJANxUAvlRAAaEBAAOuaAGdkoAIa1aEAO4AFUwkMpjANwg4AJhpDH9MZKmMAbfThNACNaYzAAawYAZW0InCgAc2QYWnRgkFcIMF8AoM1EoNoYK3CkwWM8wMyAK30ADwAhcKjY4xEAGUS4aoKtdBhtQYAmPsz42mLkEBDjEIBPf2gPbVpEmAB1NxgAC2QADgAGTTWIIM3w7Rm1uGLHXs16AEd0HHoy4wqqpD8azSCghwqXSmX0iSS-jgAEV0BB4ONNDB5ttXHtkCMkeEcP4IQBhCDCH4gKDQR4gdBBAAq83sfyCymUQA)

**Input:**
```jsx
function f() {
  return !!(
    /* comment */
    foo
  );
}
```

**Output:**
```jsx
function f() {
  return !!/* comment */
  foo;
}
```

Note that it does have the same behavior so it's not an correctness issue.

### PR DESCRIPTION
**Pull request by sosukesuzuki**
Fix #4154

Previously, Prettier removes parentheses enclose unary expressions. This change modify to keep it when the expression has comments.

```js
// Input
!!(
  /* hoge */
  foo
);

// Output(Expected)
!!(
  /* hoge */
  foo
);
```

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to the `CHANGELOG.unreleased.md` file following the template.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/printer-estree.js [hunk 1]
```diff
@@ -1630,7 +1630,28 @@ function printPathNoParens(path, options, print, args) {
         parts.push(" ");
       }
 
-      parts.push(path.call(print, "argument"));
+      if (n.argument.comments) {
+        const shouldBreak = hasNewlineInRange(
```
Comment (duailibe) [inline src/language-js/printer-estree.js]: Why is this necessary? Is this too bad?

```js
// input
!!(
  /* foo */
  foo
);

// Output
!!(/* foo */ foo);
```

Comment (sosukesuzuki) [reply src/language-js/printer-estree.js]: Sorry, I don't familiar the Prettier philosophy yet...
(Also, I don't think it is too bad.)

Should the group isn't broken?

Comment (duailibe) [reply src/language-js/printer-estree.js]: @sosukesuzuki We just try not to read from the original source as much as possible to make these kind of decisions

In this case, just ignore if the source has a new line or not

Comment (sosukesuzuki) [reply src/language-js/printer-estree.js]: I see.
For example, should Prettier how format when receive the below code.

```js
!!(
  foo
  // comment
);
```

Comment (duailibe) [reply src/language-js/printer-estree.js]: > For example, should Prettier how format when receive the below code?
> 
> ```js
> !!(
>   foo
>   // comment
> );
> ```

The same way as the input. But Prettier will do that when it prints the single line comment

Comment (sosukesuzuki) [reply src/language-js/printer-estree.js]: Okay, I fixed following it.

**Comment by sosukesuzuki** [comment]
I fixed conflicts.
Why the syntax error happens on CI...? It doesn't happen in my local env.:cry:

**Comment by sosukesuzuki** [comment]
@duailibe I consider to fix the failing tests on Azure Pipelines, but I didn't know the way to fix.
Would you please delete cache and rerun tests?

(I am not familiar with Azure Pipelines, so I do not know if it is the right way.)

**Comment by j-f1** [comment]
Looks like some tests are still failng @sosukesuzuki.

**Comment by duailibe** [comment]
@sosukesuzuki I’m on a holiday for the next couple of weeks and I’ll review this properly as soon as I get back
